File: Android/app/src/main/java/com/dreamproject/inwords/data/source/WebService/SessionHelper.java
package com.dreamproject.inwords.data.source.WebService;

import java.util.concurrent.atomic.AtomicInteger;

import javax.inject.Inject;

import io.reactivex.Completable;
import retrofit2.HttpException;

public class SessionHelper {
    private final int MAX_UNAUTHORISED_REQUESTS;
    private AtomicInteger unauthorisedReqThreshold;

    @Inject
    public SessionHelper(/*int maxRequests*/) {
        this.MAX_UNAUTHORISED_REQUESTS = 3;
        this.unauthorisedReqThreshold = new AtomicInteger(0);
    }

    public Completable interceptError(Throwable throwable) {
        if (throwable instanceof HttpException) {
            switch (((HttpException) throwable).code()) {
                case 401: //TODO find the const
                    unauthorisedReqThreshold.getAndIncrement();
                    return requireThreshold();
                default:
                    break;
            }
        }

        return Completable.complete();
    }

    public Completable requireThreshold() {
        return Completable.create(emitter -> {
            if (unauthorisedReqThreshold.get() > MAX_UNAUTHORISED_REQUESTS)
                emitter.onError(new RuntimeException("Threshold reached")); //TODO make normal exception

            emitter.onComplete();
        });
    }

    public Completable resetThreshold() {
        return Completable.fromAction(() -> unauthorisedReqThreshold.set(0));
    }
}

File: Android/app/src/main/java/ru/inwords/inwords/data/source/webService/session/SessionHelper.kt
package ru.inwords.inwords.data.source.webService.session

import retrofit2.HttpException
import java.util.concurrent.atomic.AtomicInteger
import javax.inject.Inject

class SessionHelper @Inject
internal constructor(/*int maxRequests*/) {
    companion object {
        private const val MAX_UNAUTHORISED_REQUESTS = 3
    }

    private val unauthorisedReqThreshold = AtomicInteger(0)

    fun interceptAuthError(throwable: Throwable): Boolean {
        if (throwable is HttpException) {
            when (throwable.code()) {
                401 -> { //TODO find the const
                    unauthorisedReqThreshold.getAndIncrement()
                    return true
                }
                else -> {
                }
            }
        }

        return false
    }

    fun requireThreshold() {
        if (unauthorisedReqThreshold.get() > MAX_UNAUTHORISED_REQUESTS) {
            throw RuntimeException("Threshold reached") //TODO make normal exception
        }
    }

    fun resetThreshold() {
        unauthorisedReqThreshold.set(0)
    }
}
